kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/Parcel.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.inband;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.internals.RecordHeader;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.utils.ByteUtils;
+import org.apache.kafka.common.utils.Utils;
+
+import io.kroxylicious.filter.encryption.EncryptionException;
+import io.kroxylicious.filter.encryption.ParcelVersion;
+import io.kroxylicious.filter.encryption.Receiver;
+import io.kroxylicious.filter.encryption.RecordField;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+public class Parcel {
+
+    public static final int NULL_MARKER = -1;
+    public static final int ABSENT_MARKER = -2;
+
+    private static final Header[] ABSENT_HEADERS = new Header[0];
+    private static final ByteBuffer ABSENT_VALUE = ByteBuffer.allocate(0);
+
+    static int sizeOfParcel(ParcelVersion parcelVersion, Set<RecordField> recordFields, Record kafkaRecord) {
+        switch (parcelVersion) {
+            case V1:
+                int size = sizeOfRecordValue(recordFields, kafkaRecord);
+                size += sizeOfHeaders(recordFields, kafkaRecord);
+
+                return size;
+            default:
+                throw new EncryptionException("Unknown parcel version " + parcelVersion);
+        }
+    }
+
+    static void writeParcel(ParcelVersion parcelVersion, Set<RecordField> recordFields, Record kafkaRecord, ByteBuffer parcel) {
+        switch (parcelVersion) {
+            case V1:
+                writeRecordValue(recordFields, kafkaRecord, parcel);
+                writeHeaders(recordFields, kafkaRecord, parcel);
+                return;
+            default:
+                throw new EncryptionException("Unknown parcel version " + parcelVersion);
+        }
+    }
+
+    static void readParcel(ParcelVersion parcelVersion,
+                           ByteBuffer parcel,
+                           Record encryptedRecord,
+                           @NonNull Receiver receiver) {
+        switch (parcelVersion) {
+            case V1:
+                var recordValue = readRecordValue(parcel);
+                Header[] readHeaders = readHeaders(parcel);
+                var existingHeaders = encryptedRecord.headers();
+                Header[] usedHeaders;
+                if (readHeaders == ABSENT_HEADERS) {
+                    if (existingHeaders.length > 0
+                            && InBandKeyManager.ENCRYPTION_HEADER_NAME.equals(existingHeaders[0].key())) {
```

Comment:
I guess we are leaning on the fact that we inject the `kroxylicious.io/encryption` header at the front, but I don't this it's safe to assume it will remain in the same order, maybe it could be used in conjunction with another filter that prepends another header.

We should change this to a filtering type operation that removes all encryption headers from the array.

Reply:
If there were two encryptions wouldn't it look something like:

```
headers: [kroxylicious.io/encryption=v1]
value: wrapper{parcel{headers: [kroxylicious.io/encryption=v1], value: wrapper{parcel{value: original}}}}
```

I'm wondering if encountering multiple encryption headers in the same array is an error condition.

Reply:
Your right, it would still be nested in the next payload. So yes multiple headers keyed with `kroxylicious.io/encryption` sounds wrong and highly problematic if they point at different versions. 

I'm not convinced by any strategy at this point. Remove the first header seems brittle. Removing all copies feels like it breaks workflows or masks errors, removing only the first occurrence risks passing errors along and doesn't validate versions make sense.

Reply:
added issue #813